cmd/consumer/main.go

```diff
@@ -0,0 +1,179 @@
+package main
+
+import (
+	"encoding/base64"
+	"fmt"
+	"log"
+	"math/big"
+	"os"
+	"time"
+
+	"github.com/Clever/amazon-kinesis-client-go/kcl"
+
+	"github.com/Clever/kinesis-to-firehose/firehose"
+)
+
+type RecordProcessor struct {
+	shardID           string
+	sleepDuration     time.Duration
+	checkpointRetries int
+	checkpointFreq    time.Duration
+	largestSeq        *big.Int
+	largestSubSeq     int
+	lastCheckpoint    time.Time
+	firehoseWriter    *firehose.FirehoseWriter
+}
+
+func New() *RecordProcessor {
+	return &RecordProcessor{
+		sleepDuration:     5 * time.Second,
+		checkpointRetries: 5,
+		checkpointFreq:    60 * time.Second,
+	}
+}
+
+func (rp *RecordProcessor) Initialize(shardID string) error {
+	rp.shardID = shardID
+	rp.lastCheckpoint = time.Now()
+	return nil
+}
+
+func (rp *RecordProcessor) checkpoint(checkpointer kcl.Checkpointer, sequenceNumber string, subSequenceNumber int) {
+	var err error
+	for n := 0; n < rp.checkpointRetries; n++ {
+		if err = checkpointer.Checkpoint(sequenceNumber, subSequenceNumber); err == nil {
+			return
+		}
+
+		if cperr, ok := err.(kcl.CheckpointError); ok {
+			switch cperr.Error() {
+			case "ShutdownException":
+				fmt.Fprintf(os.Stderr, "Encountered shutdown exception, skipping checkpoint\n")
+				return
+			case "ThrottlingException":
+				fmt.Fprintf(os.Stderr, "Was throttled while checkpointing, will attempt again in %s", rp.sleepDuration)
+			case "InvalidStateException":
+				fmt.Fprintf(os.Stderr, "MultiLangDaemon reported an invalid state while checkpointing\n")
+			default:
+				fmt.Fprintf(os.Stderr, "Encountered an error while checkpointing: %s", err)
+			}
+		}
+
+		if n == rp.checkpointRetries-1 {
+			fmt.Fprintf(os.Stderr, "Failed to checkpoint after %d attempts, giving up.\n", rp.checkpointRetries)
+			return
+		}
+
+		time.Sleep(rp.sleepDuration)
+	}
+}
+
+func (rp *RecordProcessor) shouldUpdateSequence(sequenceNumber *big.Int, subSequenceNumber int) bool {
+	return rp.largestSeq == nil || sequenceNumber.Cmp(rp.largestSeq) == 1 ||
```

Comment:
Added a comment 2ba6b8716600248bdf78a9057a80a5c9ee00b0a8

also, here's more info on how sequencing within the kinesis stream works: http://docs.aws.amazon.com/streams/latest/dev/kinesis-kpl-consumer-deaggregation.html